geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/features/ArrowSimpleFeature.scala

```diff
@@ -0,0 +1,122 @@
+
+package org.locationtech.geomesa.arrow.features
+
+import java.util.{Objects, Collection => jCollection, List => jList, Map => jMap}
+
+import com.vividsolutions.jts.geom.Geometry
+import org.geotools.geometry.jts.ReferencedEnvelope
+import org.locationtech.geomesa.arrow.vector.ArrowAttributeReader
+import org.locationtech.geomesa.utils.geotools.ImmutableFeatureId
+import org.opengis.feature.`type`.Name
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.feature.{GeometryAttribute, Property}
+import org.opengis.filter.identity.FeatureId
+import org.opengis.geometry.BoundingBox
+
+/**
+  * Simple feature backed by an arrow vector. Attributes are lazily evaluated - this allows filters to only
+  * examine the relevant arrow vectors for optimized reads, but also means that they are tied to the underlying
+  * vectors
+  *
+  * @param sft simple feature type
+  * @param idReader id reader
+  * @param attributeReaders attribute readers
+  * @param index index of the feature in the arrow vector
+  */
+class ArrowSimpleFeature(sft: SimpleFeatureType,
+                         idReader: ArrowAttributeReader,
+                         attributeReaders: Array[ArrowAttributeReader],
+                         private [arrow] var index: Int) extends SimpleFeature {
+  import scala.collection.JavaConversions._
+
+  private lazy val geomIndex = sft.indexOf(sft.getGeometryDescriptor.getLocalName)
+
+  override def getAttribute(i: Int): AnyRef = attributeReaders(i).apply(index)
+
+  /**
+    * Gets the underlying arrow reader for this feature
+    *
+    * @param i attribute to be read
+    * @return reader
+    */
+  def getReader(i: Int): ArrowAttributeReader = attributeReaders(i)
+
+  /**
+    * Gets the index of this feature, for use with the attribute reader
+    *
+    * @return
+    */
+  def getIndex: Int = index
+
+  override def getID: String = idReader.apply(index).asInstanceOf[String]
+  override def getIdentifier: FeatureId = new ImmutableFeatureId(getID)
+
+  override def getUserData: jMap[AnyRef, AnyRef] = Map.empty[AnyRef, AnyRef]
+
+  override def getType: SimpleFeatureType = sft
+  override def getFeatureType: SimpleFeatureType = sft
+  override def getName: Name = sft.getName
+
+  override def getAttribute(name: Name): AnyRef = getAttribute(name.getLocalPart)
+  override def getAttribute(name: String): Object = {
+    val index = sft.indexOf(name)
+    if (index == -1) { null } else { getAttribute(index) }
+  }
+
+  override def getDefaultGeometry: AnyRef = getAttribute(geomIndex)
+  override def getAttributeCount: Int = sft.getAttributeCount
+
+  override def getBounds: BoundingBox = getDefaultGeometry match {
+    case g: Geometry => new ReferencedEnvelope(g.getEnvelopeInternal, sft.getCoordinateReferenceSystem)
+    case _           => new ReferencedEnvelope(sft.getCoordinateReferenceSystem)
+  }
+
+  override def getAttributes: jList[AnyRef] = {
```

Comment:
oh i gotcha - a simple feature instance is re-used though, and `index` gets updated, so we'd have to cache by `index`